File: src/chapters/interpolation/content/code/newton-eval.cpp
#include <vector>
#include <iostream>
using namespace std;

// Evaluate the Newton form by nested (Horner-like) multiplication.
double newton_eval(const vector<double>& x, const vector<double>& a, double t) {
    double p = a.back();
    for (int k = (int)a.size() - 2; k >= 0; --k) p = p * (t - x[k]) + a[k];
    return p;
}

int main() {
    cout << newton_eval({-1, 1, 2, 3}, {-3, 2, 0, 3}, 0) << "\n";   // 5
}

File: src/chapters/interpolation/content/code/newton-eval.rs
// Evaluate the Newton form by nested (Horner-like) multiplication.
fn newton_eval(x: &[f64], a: &[f64], t: f64) -> f64 {
    let mut p = *a.last().unwrap();
    for k in (0..a.len() - 1).rev() {
        p = p * (t - x[k]) + a[k];
    }
    p
}
fn main() {
    println!("{}", newton_eval(&[-1.0, 1.0, 2.0, 3.0], &[-3.0, 2.0, 0.0, 3.0], 0.0));   // 5
}
